### ISSUE TITLE
Previous flash messages getting clobbered

### ISSUE THREAD

**Issue by aleemb** [laracasts/flash#128]
```
function update($id) {
    flash('foo');
    return redirect()->back();   
}

function edit($id) {
    flash('bar');
}
```

In the above scenario both `foo` and `bar` should be flashed. However, `flash('bar')` clobbers any previous flash messages.

**Comment by aleemb** [comment]
Pull Request: https://github.com/laracasts/flash/pull/129

**Comment by kevinschenkers** [comment]
Could this be merged? Seems to solve the issue in a nice way.

**Comment by matthew-muscat** [comment]
@JeffreyWay — are we able to get this merged? I've just ran into this behaviour on on our of projects

**Comment by axyr** [comment]
flash() is intended to show a message 1 time in the next request.
Why would you want to flash a message in the current edit view request?

**Comment by CiaranOMara** [comment]
@axyr, is correct. You may be wanting to `$request->session()->reflash();` between redirects/requests.

**Comment by cawecoy** [comment]
@CiaranOMara I don't understand how can I use `reflash()` to preserve previous flash messages from being clobbered by new ones after redirects. Can you show us how to apply your code `$request->session()->reflash();` at the @aleemb 's code to solve our issue?

**Comment by cawecoy** [comment]
@CiaranOMara no, the session is not the issue itself. `FlashNotifier`'s constructor already load the current session, but it doesn't use our current session to load the current messages flashed just now before the redirection... as a redirection implies to construct a new `FlashNotifier`, the current flashed messages are lost. As you can see in @aleemb 's example code, after a post to `update` there is a `flash()` and then a redirection to `edit` where there is another `flash()`... and when this second `flash()` is called, a new instance of `FlashNotifier` is created (**only because of the redirection**), loosing the content of the first `flash()`. `FlashNotifier`'s constructor should've get the very current messages before the redirection as well, cause it is part of the current session... as @aleemb proposes on https://github.com/laracasts/flash/pull/129.

**Comment by cawecoy** [comment]
For now, while https://github.com/laracasts/flash/pull/129 is not officially merged, there is a workaround to prevent the lost of messages flashed before redirections, for anyone who is looking for a quick solution for this issue, include the following `foreach` code just before the first `flash()` in any methods called on redirections in which you want to preserve previous `flash()`, like:

```
function update($id) {
    flash('foo');
    return redirect()->back(); // redirect(route('edit'));
}

function edit($id) {
    foreach (session('flash_notification', collect())->toArray() as $message){
        flash()->message($message);
    }
    flash('bar');
}
```

**Comment by CiaranOMara** [comment]
Going with @aleemb 's example code, I think either of the following would work.

```php
function edit(Request $request, $id) {
    $request->session()->reflash(); // using automatic injection.
    flash('bar'); 
}
```

```php
function edit($id) {
    session()->reflash(); // using the global.
    flash('bar'); 
}
```

```php
function edit($id) {
    request()->session()->reflash(); // using the global.
    flash('bar'); 
}
```

If you don't want to reflash everything, [keep](https://laravel.com/docs/5.8/session#flash-data) may be more suitable.
```php
function edit(Request $request, $id) {
    $request->session()->keep(['flash_notification']);
    flash('bar'); 
}
```

Note: sometimes the session is not inclued on the request. For example, in the context of `api` middleware.

**Comment by CiaranOMara** [comment]
The flash data gets shifted from `_flash.new` to `_flash.old` by `Illuminate\Session\Store->ageFlashData()`, which appears to be only called by `Illuminate\Session\Store->save()`. This `save` function is only called in a limited number of cases. Primarily, `save` is called by `saveSession` in the `Illuminate\Session\Middleware\StartSession` middleware. This means that the session's flash data is shifted every time a request is put through the middleware chain.

**Comment by cawecoy** [comment]
> 
> 
> Going with @aleemb 's example code, I think either of the following would work.
> 
> ```
> function edit(Request $request, $id) {
>     $request->session()->reflash(); // using automatic injection.
>     flash('bar'); 
> }
> ```
> 
> ```
> function edit($id) {
>     session()->reflash(); // using the global.
>     flash('bar'); 
> }
> ```
> 
> ```
> function edit($id) {
>     request()->session()->reflash(); // using the global.
>     flash('bar'); 
> }
> ```
> 
> If you don't want to reflash everything, [keep](https://laravel.com/docs/5.8/session#flash-data) may be more suitable.
> 
> ```
> function edit(Request $request, $id) {
>     $request->session()->keep(['flash_notification']);
>     flash('bar'); 
> }
> ```
> 
> Note: sometimes the session is not inclued on the request. For example, in the context of `api` middleware.

None of them worked for me. The ugly code `foreach (session('flash_notification', collect())->toArray() as $message){ flash($message); }` is the only working code for me until now.

**Comment by inbaz** [comment]
I've added a pull request. With the pull request that exists I get duplicate messages. Maybe you want to have a try include the code in the main project.

**Comment by jpswade** [comment]
Hi @JeffreyWay we're not merging in Pull Requests into this project? This seems like a pretty reasonable fix or should we begin to consider this abandonware?

If so, would you mark it as such.

Thanks.
